New Track Selection and Editing Selections

BryonLewis opened the pull request:
This is a bundling of a few issues related to adding in new tracks as well as selecting and editing tracks.

Fixes #50 
When adding a new track it wouldn't go to that new track in the Track Selection area.  Changing the offset for the TrackList utilizing `calcualteOffset` and watching the `selectedTrackId` and the `checkedTracks` allows us to properly move to the new item when it is created.  As part of this a new actual type of 'undefined' will be added.  This is needed to render the Track in the track list as well as provide an event in the Event viewer.  Once a user changes the track type the undefined category will be removed.

Fixes #90 
1. These were several issues with how the selection and track editing behaved.
2. Fixes the issue with editing overlapping annotations.  If you tried to edit and happened to click on another annotation it would cancel the edit.
3. Makes it easier to edit overlapping items by graying out all of the but the currently editing track.  This required importing `editingTrackId` into the useAnnotationLayer and useTextLayer files.
4. Prevents selection of other items while editing.  This means that the first click off of editing will confirm the edit instead of selecting a new item.
![image](https://user-images.githubusercontent.com/61746913/81410309-83801380-910e-11ea-8cd2-687d98ed76d6.png)

Fixes #68 
Small update to disable editing when the user clicks on the save icon.
Fixes #109 
Will fix the issue with trackId being 0 making it falsy which failed some checks in the code.  I am trying to force the other issues to occur but I think they may be fixed in the editing state adjustments for this version.

subdavis commented:
Looks like all the commits from https://github.com/VIAME/VIAME-Web/pull/81 are being caught here too.  A rebase from master should resolve that.

subdavis commented:
I think building components that rely on these sort of nasty computed views of the detection array is causing a lot of mental struggle.  I'd much rather have track as a real entity rather than just a "view" of the detection array.

I think it would make the code in `tracks.vue` much cleaner.

See https://github.com/VIAME/VIAME-Web/issues/96#issuecomment-628144052

BryonLewis commented:
> I think building components that rely on these sort of nasty computed views of the detection array is causing (me, at least) a lot of mental struggle. I'd much rather have track as a real entity rather than just a "view" of the detection array.
> 
> I think it would make the code in `tracks.vue` much cleaner.
> 
I agree that a more track centric view would make the organization and display of the data much easier.  I also think it would ease maintaining the 'edit' and 'selected' state for a track.

subdavis commented:
FWIW, we should probably still merge this when you're done with it.  I'll pull it and test, but I'm not going to look too closely at the code (unless you want me to).

BryonLewis commented:
> FWIW, we should probably still merge this when you're done with it. I'll pull it and test, but I'm not going to look too closely at the code (unless you want me to).

Sounds good if we are going to be changing some of the core structure anyways.

subdavis commented on client/src/components/layers/EditAnnotationLayer.vue:
```diff
@@ -146,6 +141,8 @@ export default {
             }
             this.$emit('update:geojson', geojson);
+          } else {
+            this.$emit('update:editing', false);
```
This breaks annotation.  

1. put track in editing state
2. Tap `f` or `d`.
3. Track editing state is removed.

BryonLewis replied:
I forgot about `f` and `d`, been using `left/right` and the editing state persists.

BryonLewis replied:
This has been updated with some comments and documentation to explain what is occuring.

AlmightyYakob commented on client/src/use/useTextLayer.js:
```diff
@@ -35,8 +36,15 @@ export default function useTextLayer({
       color: (data) => {
+        if (_editingTrackId !== null) {
+          if (_editingTrackId !== data.detection.track) {
+            return '#777777'; // color for other detections when editing
```
Since both this and `useAnnotationLayer` use this value, it might be good to put it into a constant, which through it's name would also help describes it's function.

subdavis commented:
BTW if this PR regresses the persistence feature on frame change, these methods can be removed and their children called directly in the template: 

https://github.com/VIAME/VIAME-Web/pull/95/files?file-filters%5B%5D=.vue#diff-f6924bf7e7a0bbb18d4e592590c3c21dL163-L176

BryonLewis commented:
Few Updates to behavior and Code

- Removed the references to `persistAnnotation` it is no longer needed with the editing state mode
- Updated annotation editing so that edits occur on mouse up instead of needing a confirmation click
- Checked and made sure that editing mode persists through frame advancement utilizing Arrow keys or `f` and `d` to advance through frames
- Edit mode is laggy, so it disables automatically on playback.
- While adding a brand new Track or a new Detection to an existing track the `escape` key will cancel out of editing/selection mode at any time.
- Refactored the `usetypeColorMap` into `useStyling` which sets up all of the annotation and type styling for the AnnotationEditor.  `selected | standard | disabled ` are different `stateStyling` consumed by the Text and AnnotationLayer.  I imagine that `editAnnotationLayer` will use stuff out of here to setup the styles for the edit handles in the future.

Future things not in this PR:

- Context sensitive deletion.  For the ability to delete an individual annotation instead of the entire track when in editing mode
- Styling for the EditHandles (The big blue circles used during editing)

subdavis commented:
1. dragging a handle makes it no longer selectable until you move your mouse off and back on again

> Prevents selection of other items while editing. This means that the first click off of editing will confirm the edit instead of selecting a new item.

2. I played with it for a bit, and I don't think I like this behavior.  I found the extra click frustrating and confusing.  I understand the issue with selection of clustered boxes, but this seemed to really degrade my experience.  (Also worth noting that this isn't how any annotation tool I'm familiar with works).

Perhaps feedback from @AlmightyYakob would be helpful here.
